gdnative-core/src/dictionary.rs

```diff
@@ -11,121 +12,164 @@ use crate::Variant;
     }
 
     /// Returns a mutable reference to the value corresponding to the key.
+    ///
+    /// # Safety
+    ///
+    /// The returned reference is invalidated if the same container is mutated through another
+    /// reference. It is possible to create two mutable references to the same memory location
+    /// if the same `key` is provided, causing undefined behavior.
     #[inline]
-    pub fn get_mut_ref(&mut self, key: &Variant) -> &mut Variant {
-        unsafe {
-            Variant::cast_mut_ref((get_api().godot_dictionary_operator_index)(
-                &mut self.0,
-                &key.0,
-            ))
-        }
+    #[allow(clippy::mut_from_ref)]
+    pub unsafe fn get_mut_ref(&self, key: &Variant) -> &mut Variant {
+        Variant::cast_mut_ref((get_api().godot_dictionary_operator_index)(
+            self.sys_mut(),
+            key.sys(),
+        ))
     }
 
     /// Returns a GodotString of the `Dictionary`.
     #[inline]
     pub fn to_json(&self) -> GodotString {
-        unsafe { GodotString((get_api().godot_dictionary_to_json)(&self.0)) }
+        unsafe { GodotString((get_api().godot_dictionary_to_json)(self.sys())) }
     }
 
     /// Returns an array of the keys in the `Dictionary`.
     #[inline]
-    pub fn keys(&self) -> VariantArray {
-        unsafe { VariantArray((get_api().godot_dictionary_keys)(&self.0)) }
+    pub fn keys(&self) -> VariantArray<Shared> {
+        unsafe { VariantArray::from_sys((get_api().godot_dictionary_keys)(self.sys())) }
     }
 
     /// Returns an array of the values in the `Dictionary`.
     #[inline]
     pub fn values(&self) -> VariantArray {
-        unsafe { VariantArray((get_api().godot_dictionary_values)(&self.0)) }
+        unsafe { VariantArray::from_sys((get_api().godot_dictionary_values)(self.sys())) }
     }
 
     #[inline]
     pub fn get_next(&self, key: &Variant) -> &Variant {
-        unsafe { Variant::cast_ref((get_api().godot_dictionary_next)(&self.0, &key.0)) }
+        unsafe { Variant::cast_ref((get_api().godot_dictionary_next)(self.sys(), key.sys())) }
     }
 
     /// Return a hashed i32 value representing the dictionary's contents.
     #[inline]
     pub fn hash(&self) -> i32 {
-        unsafe { (get_api().godot_dictionary_hash)(&self.0) }
+        unsafe { (get_api().godot_dictionary_hash)(self.sys()) }
+    }
+
+    #[doc(hidden)]
+    #[inline]
+    pub fn sys(&self) -> *const sys::godot_dictionary {
+        &self.sys
+    }
+
+    #[doc(hidden)]
+    #[inline]
+    pub fn sys_mut(&self) -> *mut sys::godot_dictionary {
+        &self.sys as *const _ as *mut _
+    }
+
+    #[doc(hidden)]
+    #[inline]
+    pub fn from_sys(sys: sys::godot_dictionary) -> Self {
+        Dictionary {
+            sys,
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// Operations allowed on Dictionaries that might be shared.
+impl Dictionary<Shared> {
+    /// Assume the dictionary is only used/referenced by a single thread.
+    ///
+    ///
```

Comment:
I feel like this might not be the best wording, but I'm not sure how it can be improved. Maybe the following?

```
Assume that this is the only reference to this dictionary, on which operations that change the container size can be safely performed.

# Safety

It isn't thread-safe to perform operations that change the container size from multiple threads at the same time. Creating multiple `Unique` references to the same collections, or violating the thread-safety guidelines in non-Rust code will cause undefined behavior.
```

It might be a bit obtuse, but I think this captures the semantics of `Unique` more accurately.

Reply:
Thanks, I will use this instead!